File: merge_sort/merge-sort.ts
export {};

function mergeSort(nums: number[]): number[] {
  if (nums.length < 2) {
    return nums;
  }

  const middle = Math.floor(nums.length / 2);
  const [sortedLeftSide, sortedRightSide] = [
    mergeSort(nums.slice(0, middle)),
    mergeSort(nums.slice(middle)),
  ];

  return merge(sortedLeftSide, sortedRightSide);
}

function merge(leftSide: number[], rightSide: number[]): number[] {
  const final: number[] = [];
  let [i, j] = [0, 0];

  while (i < leftSide.length && j < rightSide.length) {
    const [leftSideElement, rightSideElement] = [leftSide[i], rightSide[j]];
    if (leftSideElement < rightSideElement) {
      final.push(leftSideElement);
      i++;
    } else {
      final.push(rightSideElement);
      j++;
    }
  }

  while (i < leftSide.length) {
    final.push(leftSide[i]);
    i++;
  }
  while (j < rightSide.length) {
    final.push(rightSide[j]);
    j++;
  }

  return final;
}

File: merge_sort/merge-sort.go
package main

func main() {

}

func mergeSort(nums []int) []int {
	if len(nums) < 2 {
		return nums
	}
	middle := len(nums) / 2
	sortedLeftSide, sortedRightSide := mergeSort(nums[:middle]), mergeSort(nums[middle:])
	return merge(sortedLeftSide, sortedRightSide)
}

func merge(leftSide, rightSide []int) []int {
	final := []int{}
	i, j := 0, 0

	for i < len(leftSide) && j < len(rightSide) {
		leftSideElement, rightSideElement := leftSide[i], rightSide[j]
		if leftSideElement < rightSideElement {
			final = append(final, leftSideElement)
			i++
		} else {
			final = append(final, rightSideElement)
			j++
		}
	}

	if i < len(leftSide) {
		final = append(final, leftSide[i:]...)
	}
	if j < len(rightSide) {
		final = append(final, rightSide[j:]...)
	}

	return final
}
